sys/stdio_nimble/stdio_nimble.c

```diff
@@ -0,0 +1,399 @@
+{
+    /*
+     * access_cb defines a callback for read and write access events on
+     * given characteristics
+     */
+    {
+        /* Service: stdio */
+        .type = BLE_GATT_SVC_TYPE_PRIMARY,
+        .uuid = (ble_uuid_t*) &gatt_svr_svc_stdio_uuid.u,
+        .characteristics = (struct ble_gatt_chr_def[]) { {
+            /* Characteristic: stdout */
+            .uuid = (ble_uuid_t*) &gatt_svr_chr_stdout_uuid.u,
+            .access_cb = gatt_svr_chr_access_noop,
+            .val_handle = &_val_handle_stdout,
+            .flags = BLE_GATT_CHR_F_INDICATE,
+        }, {
+            /* Characteristic: stdin */
+            .uuid = (ble_uuid_t*) &gatt_svr_chr_stdin_uuid.u,
+            .access_cb = gatt_svr_chr_access_stdin,
+            .flags = BLE_GATT_CHR_F_WRITE,
+        }, {
+            0, /* No more characteristics in this service */
+        }, }
+    },
+    {
+        0, /* No more services */
+    },
+};
+
+static int _gap_event_cb(struct ble_gap_event *event, void *arg);
+
+static void _send_from_ringbuffer(bool mutex_locked);
+
+void stdio_init(void)
+{
+    ringbuffer_init(&_rb_stdin, _rb_stdin_mem, CONFIG_STDIO_NIMBLE_STDIN_BUFSIZE);
+    ringbuffer_init(&_rb_stdout, _rb_stdout_mem, CONFIG_STDIO_NIMBLE_STDOUT_BUFSIZE);
+
+    uart_init(STDIO_UART_DEV, STDIO_UART_BAUDRATE, NULL, NULL);
+}
+
+ssize_t stdio_read(void *buffer, size_t count)
+{
+    ssize_t res;
+
+    /* block until at least on character was read from the buffer */
+    while (!(res = ringbuffer_get(&_rb_stdin, buffer, count))) {};
+
+    return res;
+}
+
+ssize_t stdio_write(const void* buffer, size_t len)
+{
+    mutex_lock(&_rb_stdout_mutex);
+
+    unsigned int free = ringbuffer_get_free(&_rb_stdout);
+    unsigned int consumed = 0;
+
+    if (free > 0) {
+        unsigned int to_send = (free <= len) ? free : len;
+        consumed = ringbuffer_add(&_rb_stdout, buffer, to_send);
+    }
+
+    mutex_unlock(&_rb_stdout_mutex);
+
+    /* initially trigger chained async calls (chained via "_gap_event_cb") */
+    if (mutex_trylock(&_send_mutex)) {
+        if (!_sending_in_progress) {
+            _send_from_ringbuffer(true);
+        }
+        else {
+            /* there is an active call chain, we must not initiate a new one */
+            mutex_unlock(&_send_mutex);
+        }
+    }
+
+    return consumed;
+}
+
+/* is going to be called by auto_init */
+void stdio_nimble_init(void)
+{
+    int rc = 0;
+
+    /* verify and add our custom services */
+    rc = ble_gatts_count_cfg(_gatt_svr_svcs);
+    assert(rc == 0);
+    rc = ble_gatts_add_svcs(_gatt_svr_svcs);
+    assert(rc == 0);
+
+    /* reload the GATT server to link our added services */
+    ble_gatts_start();
+
+    /* register gap event listener */
+    rc = ble_gap_event_listener_register(&_gap_event_listener, _gap_event_cb, NULL);
+    assert(rc == 0);
+
+    /* fix compilation error when using DEVELHELP=0 */
+    (void) rc;
+}
+
+/* there should be always only one active call chain */
```

Comment:
Can you five more details on why?

Reply:
Now with the callout functions this is obsolete. Before I had to chain it the send function by the gap event callback. And a new call of `stdio_write` should not have started a second chain. Therefore `_send_mutex` is now obsolete too. :D